phoenix-core/src/it/java/org/apache/phoenix/end2end/OnDuplicateKey2IT.java

```diff
@@ -118,6 +137,148 @@ public void testIgnoreReturnValue() throws Exception {
         conn.close();
     }
 
+    @Test
+    public void testReturnRowResult() throws Exception {
+        Assume.assumeTrue("Set correct result to RegionActionResult on hbase versions " +
+                "2.4.18+, 2.5.9+, and 2.6.0+", isSetCorrectResultEnabledOnHBase());
+
+        Properties props = PropertiesUtil.deepCopy(TEST_PROPERTIES);
+        String sample1 = getJsonString("json/sample_01.json");
+        String sample2 = getJsonString("json/sample_02.json");
+        BsonDocument bsonDocument1 = RawBsonDocument.parse(sample1);
+        BsonDocument bsonDocument2 = RawBsonDocument.parse(sample2);
+        try (Connection conn = DriverManager.getConnection(getUrl(), props)) {
+            conn.setAutoCommit(true);
+            String tableName = generateUniqueName();
+            String ddl = "CREATE TABLE " + tableName
+                    + "(PK1 VARCHAR, PK2 DOUBLE NOT NULL, PK3 VARCHAR, COUNTER1 DOUBLE,"
+                    + " COUNTER2 VARCHAR,"
+                    + " COL3 BSON, COL4 INTEGER, CONSTRAINT pk PRIMARY KEY(PK1, PK2, PK3))";
+            conn.createStatement().execute(ddl);
+            createIndex(conn, tableName);
+
+            String upsertSql = "UPSERT INTO " + tableName + " (PK1, PK2, PK3, COUNTER1, COL3, COL4)"
+                    + " VALUES('pk000', -123.98, 'pk003', 1011.202, ?, 123) ON DUPLICATE KEY " +
+                    "IGNORE";
+            validateReturnedRowAfterUpsert(conn, upsertSql, tableName, 1011.202, null, true,
+                    bsonDocument1, bsonDocument1, 123);
+
+            upsertSql =
+                    "UPSERT INTO " + tableName + " (PK1, PK2, PK3, COUNTER1) "
+                            + "VALUES('pk000', -123.98, 'pk003', 0) ON DUPLICATE"
+                            + " KEY IGNORE";
+            validateReturnedRowAfterUpsert(conn, upsertSql, tableName, 1011.202, null, false,
+                    null, bsonDocument1, 123);
+
+            upsertSql =
+                    "UPSERT INTO " + tableName
+                            + " (PK1, PK2, PK3, COUNTER1, COUNTER2) VALUES('pk000', -123.98, "
+                            + "'pk003', 234, 'col2_000')";
+            validateReturnedRowAfterUpsert(conn, upsertSql, tableName, 234d, "col2_000", true,
```

Comment:
In case of partial update, will the returned row be a full row ?

Reply:
In the above example we are doing a regular upsert with returned row.

Reply:
> In case of partial update, will the returned row be a full row ?

That's correct, full row version will be returned. That's the main purpose actually: client can perform partial row update and see what the entire row looks like.

> In the above example we are doing a regular upsert with returned row.

Yes, just an example of how DELETE will also look like. But the point is, the server allows returning row only if client uses new API to modify only single row.